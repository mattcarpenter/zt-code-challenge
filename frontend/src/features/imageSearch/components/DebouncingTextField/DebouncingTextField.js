import {TextField} from '@material-ui/core'
import PropTypes from 'prop-types';
import React from 'react'

const DebouncingTextField = ({label, value, debounceTime, onChange, placeholder, className, onFocus}) => {

  const handleChange = debounceEvent((value) => {
    onChange(value);
  }, debounceTime);

  return (
    <TextField
      className={className}
      label={label}
      type="text"
      variant="outlined"
      value={value}
      onChange={handleChange}
      placeholder={placeholder}
      onFocus={onFocus}
      fullWidth={true}
    />
  );
};

DebouncingTextField.propTypes = {
  label: PropTypes.string,
  onChange: PropTypes.func,
  value: PropTypes.string,
  placeholder: PropTypes.string,
  className: PropTypes.string,
  onFocus: PropTypes.func
};

DebouncingTextField.defaultProps = {
  debounceTime: 250,
  onChange: () => {},
  onFocus: () => {}
};

export const debounceEvent = (callback, time, interval) => {
  return (event) => {
    const value = event.target.value;
    clearTimeout(interval, interval = setTimeout(() => callback(value), time));
  }
};

export default DebouncingTextField;
