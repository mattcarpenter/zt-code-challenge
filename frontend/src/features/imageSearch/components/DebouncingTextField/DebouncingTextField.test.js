import {mount, shallow} from 'enzyme';
import React from 'react';
import DebouncingTextField from './DebouncingTextField';

describe('debouncingTextBox', () => {

  it('renders properly', () => {
    const wrapper = shallow(<DebouncingTextField placeholder="Placeholder text"/>);
    expect(wrapper).toMatchSnapshot();
  });

  it('debounces user input and calls onChange prop method', done => {
    const handleChange = jest.fn();
    const wrapper = mount(
      <DebouncingTextField onChange={handleChange} />
    );

    const input = wrapper.find('input');
    input.simulate('change', { target: { value: 'a' } });
    input.simulate('change', { target: { value: 'ab' } });
    input.simulate('change', { target: { value: 'abc' } });

    setTimeout(function () {
      expect(handleChange.mock.calls.length).toBe(1);
      expect(handleChange.mock.calls[0][0]).toBe('abc');
      done();
    }, 300);
  });
});
