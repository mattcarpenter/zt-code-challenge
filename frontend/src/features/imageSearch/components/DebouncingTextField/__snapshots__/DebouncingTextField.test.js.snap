// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`debouncingTextBox renders properly 1`] = `
<WithStyles(ForwardRef(TextField))
  fullWidth={true}
  onChange={[Function]}
  onFocus={[Function]}
  placeholder="Placeholder text"
  type="text"
  variant="outlined"
/>
`;
